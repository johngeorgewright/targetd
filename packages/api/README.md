# @targetd/api

Load and query data

## Installation

| JS Runtime | Command                     |
| ---------- | --------------------------- |
| Node.js    | `npx jsr add @targetd/api`  |
| Bun        | `bunx jsr add @targetd/api` |
| Deno       | `deno add jsr:@targetd/api` |
