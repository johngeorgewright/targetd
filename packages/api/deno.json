{
  "name": "@targetd/api",
  "version": "7.1.0",
  "exports": "./src/index.ts",
  "fmt": {
    "semiColons": false,
    "singleQuote": true
  },
  "tasks": {
    "pwd": "echo $(pwd)",
    "test": "deno test --allow-all --trace-leaks"
  },
  "imports": {
    "@myty/promise-chain": "jsr:@myty/promise-chain@^0.4.6",
    "fast-deep-equal": "npm:fast-deep-equal@^3.0.0",
    "ts-toolbelt": "npm:ts-toolbelt@^9.0.0",
    "zod": "npm:zod@^4.0.15"
  },
  "publish": {
    "include": [
      "deno.json",
      "src/",
      "LICENSE",
      "README.md"
    ]
  }
}
