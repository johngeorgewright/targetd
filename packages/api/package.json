{
  "name": "@targetd/api",
  "version": "6.0.1",
  "description": "Load and query data",
  "main": "dist/index.js",
  "sideEffects": false,
  "scripts": {
    "build": "yarn clean && yarn tsc",
    "clean": "yarn rimraf dist",
    "start": "yarn tsc --watch --preserveWatchOutput",
    "test": "yarn jest --passWithNoTests"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/johngeorgewright/config/issues"
  },
  "homepage": "https://github.com/johngeorgewright/config#readme",
  "devDependencies": {
    "@semantic-release/commit-analyzer": "11.1.0",
    "@semantic-release/exec": "6.0.3",
    "@semantic-release/git": "10.0.1",
    "@semantic-release/github": "9.2.6",
    "@semantic-release/release-notes-generator": "12.1.0",
    "@types/jest": "29.5.13",
    "@types/node": "20.16.9",
    "jest": "29.7.0",
    "prettier-2": "npm:prettier@^2",
    "rimraf": "6.0.1",
    "semantic-release": "23.1.1",
    "semantic-release-monorepo": "8.0.2",
    "ts-jest": "29.2.5",
    "typescript": "5.4.5",
    "zod": "3.23.8"
  },
  "dependencies": {
    "fast-deep-equal": "3.1.3",
    "ts-toolbelt": "9.6.0",
    "tslib": "2.7.0"
  },
  "peerDependencies": {
    "zod": "^3.20.0"
  }
}
