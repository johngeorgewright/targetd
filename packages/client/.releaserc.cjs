// @ts-check

/**
 * @type {import('npm:semantic-release').Options}
 */
module.exports = {
  extends: [
    '../../.releaserc.cjs',
    'semantic-release-monorepo',
  ],
}
