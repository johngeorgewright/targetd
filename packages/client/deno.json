{
  "name": "@targetd/client",
  "version": "6.0.0",
  "exports": "./src/index.ts",
  "imports": {
    "@targetd/api": "jsr:@targetd/api@^7.0.0",
    "ts-toolbelt": "npm:ts-toolbelt@^9.6.0",
    "zod": "npm:zod@^4.0.15"
  },
  "fmt": {
    "semiColons": false,
    "singleQuote": true
  },
  "tasks": {
    "pwd": "echo $(pwd)",
    "test": "deno test --allow-all --trace-leaks"
  },
  "publish": {
    "include": [
      "deno.json",
      "src/",
      "LICENSE",
      "README.md"
    ]
  }
}
