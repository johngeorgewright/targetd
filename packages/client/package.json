{
  "name": "@targetd/client",
  "version": "5.0.0",
  "description": "Typed querying of a @targetd/server",
  "main": "dist/index.js",
  "sideEffects": false,
  "scripts": {
    "build": "yarn clean && yarn tsc && yarn rimraf 'dist/**/?(__tests__|__mocks__|__setup__|*.test.*)'",
    "clean": "yarn rimraf dist",
    "start": "yarn tsc --watch --preserveWatchOutput",
    "release": "yarn semantic-release -e semantic-release-monorepo",
    "test": "yarn jest --passWithNoTests"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/johngeorgewright/config/issues"
  },
  "homepage": "https://github.com/johngeorgewright/config#readme",
  "devDependencies": {
    "@semantic-release/commit-analyzer": "11.1.0",
    "@semantic-release/exec": "6.0.3",
    "@semantic-release/git": "10.0.1",
    "@semantic-release/github": "9.2.6",
    "@semantic-release/release-notes-generator": "12.1.0",
    "@targetd/api": "workspace:*",
    "@targetd/date-range": "workspace:*",
    "@targetd/server": "workspace:*",
    "@types/jest": "29.5.13",
    "@types/node": "20.16.9",
    "express": "4.18.3",
    "jest": "29.7.0",
    "rimraf": "6.0.1",
    "semantic-release": "23.1.1",
    "semantic-release-monorepo": "8.0.2",
    "ts-jest": "29.2.5",
    "typescript": "5.4.5",
    "zod": "3.23.8"
  },
  "dependencies": {
    "@types/express": "4.17.21",
    "ts-toolbelt": "9.6.0",
    "tslib": "2.7.0"
  },
  "peerDependencies": {
    "@targetd/api": ">=6.0.0",
    "zod": "^3.20.0"
  }
}
