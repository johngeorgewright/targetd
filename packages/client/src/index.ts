import type { Data, DT, PT, StaticRecord } from '@targetd/api'
import type { $ZodShape } from 'zod/v4/core'
import { queryToURLSearchParams } from './queryToURLSearchParams.ts'
import { ZodError } from 'zod'
import { ResponseError } from './ResponseError.ts'

export class Client<
  PayloadParsers extends $ZodShape,
  TargetingParsers extends $ZodShape,
  QueryParsers extends $ZodShape,
  FallThroughTargetingParsers extends $ZodShape,
> {
  #baseURL: string

  #data: Data<
    PayloadParsers,
    TargetingParsers,
    QueryParsers,
    FallThroughTargetingParsers
  >

  #init?: RequestInit

  constructor(
    baseURL: string,
    data: Data<
      PayloadParsers,
      TargetingParsers,
      QueryParsers,
      FallThroughTargetingParsers
    >,
    init?: RequestInit,
  ) {
    this.#baseURL = baseURL
    this.#data = data.removeAllRules()
    this.#init = init
  }

  async getPayload<Name extends keyof PayloadParsers>(
    name: Name,
    rawQuery: Partial<StaticRecord<QueryParsers>> = {},
  ): Promise<
    PT.Payload<PayloadParsers[Name], FallThroughTargetingParsers> | void
  > {
    const query = this.#data.QueryParser.parse(rawQuery)
    const urlSearchParams = queryToURLSearchParams(query)
    const response = await fetch(
      `${this.#baseURL}/${String(name)}?${urlSearchParams}`,
      {
        method: 'GET',
        ...this.#init,
      },
    )

    switch (true) {
      case response.status === 204:
        return undefined
      case response.status === 400:
        await response.json()
          .then(
            (error) => {
              if (error.name === '$ZodError') {
                throw new ZodError(JSON.parse(error.message))
              }
            },
            () => {},
          )
      // fallthrough
      case response.status > 200 || response.status < 200:
        throw new ResponseError(response)
      default: {
        const data = await this.#data.insert({
          [name]: await response.json(),
        } as any)
        return data.getPayload(name)
      }
    }
  }

  async getPayloadForEachName(
    rawQuery: Partial<StaticRecord<QueryParsers>> = {},
  ): Promise<
    Partial<
      {
        [Name in keyof PayloadParsers]:
          | PT.Payload<PayloadParsers[Name], FallThroughTargetingParsers>
          | undefined
      }
    >
  > {
    const query = this.#data.QueryParser.parse(rawQuery)
    const urlSearchParams = queryToURLSearchParams(query)
    const response = await fetch(`${this.#baseURL}?${urlSearchParams}`, {
      method: 'GET',
      ...this.#init,
    })
    const data = await this.#data.insert((await response.json()) as any)
    return data.getPayloadForEachName()
  }
}

export type ClientWithData<
  D extends DT.Any,
> = Client<
  DT.PayloadParsers<D>,
  DT.TargetingParsers<D>,
  DT.QueryParsers<D>,
  DT.FallThroughTargetingParsers<D>
>
