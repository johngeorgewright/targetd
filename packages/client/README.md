# @targetd/client

> A HTTP typed client to query a @targetd/server

## Installation

| JS Runtime | Command                                     |
| ---------- | ------------------------------------------- |
| Node.js    | `npx jsr add @targetd/api @targetd/client`  |
| Bun        | `bunx jsr add @targetd/api @targetd/client` |
| Deno       | `deno add @targetd/api @targetd/client`     |

## Example

```typescript
// ./device.ts
import { createTargetingDescriptor } from '@targetd/api'
import { z } from 'zod'

export const Device = z.literal('desktop').or(z.literal('mobile'))

export const deviceTargeting = createTargetingDescriptor({
  predicate: (q) => (t) => typeof q === 'string' && t.includes(q),
  queryParser: Device,
  targetingParser: z.array(Device),
})
```

```typescript
// ./data.ts
import { Data } from '@targetd/api'
import z from 'zod'
import { deviceTargeting } from './device'

export const data = Data.create({
  data: {
    bar: z.number(),
    foo: z.string(),
  },
  targeting: {
    device: deviceTargeting,
  },
})
```

```typescript
// ./server.ts
import { createServer } from '@targetd/server'
import { data } from './data'

createServer(
  data
    .addRules('bar', [
      {
        payload: 123,
      },
    ])
    .addRules('foo', [
      {
        targeting: {
          device: ['mobile'],
        },
        payload: '‚Äçüì±',
      },
      {
        targeting: {
          device: ['desktop'],
        },
        payload: 'üñ•',
      },
    ]),
).listen(3_000)
```

```typescript
// ./client.ts
import { Client } from '@targetd/client'
import z from 'zod'
import { data } from './data'

const client = new Client('http://localhost:3000', data)

expect(await clientData.getPayloadForEachName({ device: 'mobile' })).toEqual({
  bar: 123,
  foo: '‚Äçüì±',
})

expect(await clientData.getPayloadForEachName({ device: 'desktop' })).toEqual({
  bar: 123,
  foo: 'üñ•',
})
```
