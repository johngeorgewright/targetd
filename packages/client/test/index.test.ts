import { assertSnapshot } from 'jsr:@std/testing/snapshot'
import { afterAll, beforeAll, test } from 'jsr:@std/testing/bdd'
import { expect } from 'jsr:@std/expect'
import { Data, targetEquals, targetIncludes } from '@targetd/api'
import dateRangeTargeting from '@targetd/date-range'
import { createServer } from '@targetd/server'
import type { Server } from 'node:http'
import type { AddressInfo } from 'node:net'
import { setTimeout } from 'node:timers/promises'
import z from 'zod/v4'
import { Client, type ClientWithData } from '@targetd/client'

const schema = Data.create({
  data: {
    foo: z.string(),
    bar: z.number(),
    timed: z.string(),
  },
  targeting: {
    weather: targetIncludes(z.string()),
    highTide: targetEquals(z.boolean()),
    asyncThing: {
      predicate: (q) => setTimeout(10, (t) => q === t && setTimeout(10, true)),
      queryParser: z.boolean(),
      targetingParser: z.boolean(),
    },
    date: dateRangeTargeting,
  },
})

const data = schema
  .addRules('foo', [
    {
      targeting: {
        highTide: true,
        weather: ['sunny'],
      },
      payload: '🏄‍♂️',
    },
    {
      targeting: {
        weather: ['sunny'],
      },
      payload: '😎',
    },
    {
      targeting: {
        weather: ['rainy'],
      },
      payload: '☂️',
    },
    {
      targeting: {
        highTide: true,
      },
      payload: '🌊',
    },
    {
      targeting: {
        asyncThing: true,
      },
      payload: 'Async payload',
    },
    {
      payload: 'bar',
    },
  ])
  .then((data) =>
    data.addRules('bar', [
      {
        payload: 123,
      },
    ])
  )
  .then((data) =>
    data.addRules('timed', [
      {
        targeting: {
          date: { start: '2001-01-01', end: '2010-01-01' },
        },
        payload: 'in time',
      },
    ])
  )

let client: ClientWithData<Awaited<typeof schema>>
let server: Server

beforeAll(async () => {
  const d = await data
  const serverResolvers = Promise.withResolvers<void>()
  server = createServer(() => d).listen(
    0,
    serverResolvers.resolve,
  )
  await serverResolvers.promise
  const address = server.address() as AddressInfo
  client = new Client(`http://localhost:${address.port}`, schema)
})

afterAll(() => {
  server.close()
})

test('get one data point', async (t) => {
  expect(await client.getPayload('foo')).toBe('bar')
  expect(await client.getPayload('foo', { weather: 'sunny' })).toBe('😎')
  expect(await client.getPayload('foo', { weather: 'rainy' })).toBe('☂️')
  expect(await client.getPayload('foo', { highTide: true })).toBe('🌊')
  expect(
    await client.getPayload('foo', { highTide: true, weather: 'sunny' }),
  ).toBe('🏄‍♂️')
  await assertSnapshot(t, await client.getPayload('foo', { asyncThing: true }))
  await assertSnapshot(
    t,
    await client.getPayload('timed', { date: { start: '2002-01-01' } }),
  )
  expect(
    await client.getPayload('timed', { date: { start: '2012-01-01' } }),
  ).toBe(undefined)
})

test('get all', async (t) => {
  await assertSnapshot(t, await client.getPayloadForEachName())
  await assertSnapshot(
    t,
    await client.getPayloadForEachName({ weather: 'sunny' }),
  )
  await assertSnapshot(
    t,
    await client.getPayloadForEachName({ weather: 'rainy' }),
  )
  await assertSnapshot(
    t,
    await client.getPayloadForEachName({ highTide: true }),
  )
  await assertSnapshot(
    t,
    await client.getPayloadForEachName({ highTide: true, weather: 'sunny' }),
  )
  await assertSnapshot(
    t,
    await client.getPayloadForEachName({ asyncThing: true }),
  )
})
