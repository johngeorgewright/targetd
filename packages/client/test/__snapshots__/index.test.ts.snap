export const snapshot = {};

snapshot[`global > get one data point 1`] = `"Async payload"`;

snapshot[`global > get one data point 2`] = `"in time"`;

snapshot[`global > get all 1`] = `
{
  bar: 123,
  foo: "bar",
}
`;

snapshot[`global > get all 2`] = `
{
  bar: 123,
  foo: "😎",
}
`;

snapshot[`global > get all 3`] = `
{
  bar: 123,
  foo: "☂️",
}
`;

snapshot[`global > get all 4`] = `
{
  bar: 123,
  foo: "🌊",
}
`;

snapshot[`global > get all 5`] = `
{
  bar: 123,
  foo: "🏄‍♂️",
}
`;

snapshot[`global > get all 6`] = `
{
  bar: 123,
  foo: "Async payload",
}
`;
